Stripping trailing zeros stops round trip capabality for models.

This now actually fails one of my real data tests 

FAILED tests/model/test_ech0196.py::test_xml_round_trip_files[/workspaces/opensteuerauszug/private/samples/UBS_fixed.xml]

because apparently UBS does not do this. Need to ponder what to do here..

_Originally posted by @vroonhof in https://github.com/vroonhof/opensteuerauszug/issues/144#issuecomment-3906064170_

--- a/src/opensteuerauszug/model/ech0196.py
+++ b/src/opensteuerauszug/model/ech0196.py
@@ -421,8 +421,8 @@ class BaseXmlModel(BaseModel):
                         # TODO: Implement base64 encoding if needed for fileData
                         raise NotImplementedError("Base64 encoding is not implemented")
                     elif isinstance(value, Decimal):
-                        # Format Decimal as a plain string to avoid scientific notation
-                        str_value = f'{value:f}'
+                        # Format Decimal as a plain string to avoid scientific notation and remove trailing zeros
+                        str_value = format(value.normalize(), "f")
                     else:
                         str_value = str(value)
                     element.set(attr_name, str_value)

--- a/tests/model/test_ech0196.py
+++ b/tests/model/test_ech0196.py
@@ -51,10 +51,10 @@ def sample_tax_statement_data():
                 salutation="2"
             )
         ],
-        totalTaxValue=Decimal("1000.50"),
-        totalGrossRevenueA=Decimal("100.00"),
-        totalGrossRevenueB=Decimal("50.00"),
-        totalWithHoldingTaxClaim=Decimal("35.00")
+        totalTaxValue=Decimal("1000.5"),
+        totalGrossRevenueA=Decimal("100"),
+        totalGrossRevenueB=Decimal("50"),
+        totalWithHoldingTaxClaim=Decimal("35")
     )
 
 # Use the centralized helper function
@@ -74,7 +74,7 @@ def test_tax_statement_creation(sample_tax_statement_data):
     assert statement.institution.name == "Test Bank AG"
     assert len(statement.client) == 1
     assert statement.client[0].lastName == "Muster"
-    assert statement.totalTaxValue == Decimal("1000.50")
+    assert statement.totalTaxValue == Decimal("1000.5")
 
 def test_tax_statement_to_xml(sample_tax_statement_data):
     """Tests serialization to XML bytes and checks basic structure."""
@@ -91,7 +91,7 @@ def test_tax_statement_to_xml(sample_tax_statement_data):
         assert root.get('minorVersion') == "2"
         assert root.get('id') == "test-id-123"
         assert root.get('canton') == "ZH"
-        assert root.get('totalTaxValue') == "1000.50"
+        assert root.get('totalTaxValue') == "1000.5"
         # Check institution element existence and name attribute
         institution_el = root.find(ns_tag('eCH-0196', 'institution'), namespaces=NS_MAP)
         assert institution_el is not None
@@ -190,7 +190,7 @@ def test_bank_account_round_trip():
                  bankAccountNumber="ACC987"
                  bankAccountName="Main Account">
       <taxValue referenceDate="2024-12-31" balanceCurrency="CHF" balance="10000" exchangeRate="1" value="10000"/>
-      <payment paymentDate="2024-03-31" name="Habenzins mit Verrechnungssteuer" amountCurrency="CHF" amount="1.20" exchangeRate="1" grossRevenueA="1.20" grossRevenueB="0" withHoldingTaxClaim=".3"/>
+      <payment paymentDate="2024-03-31" name="Habenzins mit Verrechnungssteuer" amountCurrency="CHF" amount="1.2" exchangeRate="1" grossRevenueA="1.2" grossRevenueB="0" withHoldingTaxClaim=".3"/>
     </bankAccount>
     '''
     parser = ET.XMLParser(remove_blank_text=True)
@@ -511,6 +511,59 @@ def test_decimal_serialization_avoids_scientific_notation():
     assert serialized_value is not None, "totalTaxValue attribute should be serialized"
     assert "E" not in serialized_value.upper(), f"Scientific notation found in serialized decimal: {serialized_value}"
     assert serialized_value == "0.0000000001", f"Serialized decimal has incorrect format: {serialized_value}"
+
+def test_decimal_trailing_zero_removal():
+    """Test that trailing zeros are removed from decimal serialization in XML output."""
+    from decimal import Decimal
+    from opensteuerauszug.model.ech0196 import BankAccount, BankAccountName, BankAccountNumber
+    import lxml.etree as ET
+
+    temp_parent = ET.Element("temp", nsmap={None: 'http://www.ech.ch/xmlns/eCH-0196/2'})
+    # 0.000 -> 0
+    ba1 = BankAccount(
+        iban="CH0",
+        bankAccountNumber=BankAccountNumber("ACC0"),
+        bankAccountName=BankAccountName("Test"),
+        totalTaxValue=Decimal("0.000")
+    )
+    el1 = ba1._build_xml_element(temp_parent)
+    assert el1.get("totalTaxValue") == "0"
+    # 1.2300 -> 1.23
+    ba2 = BankAccount(
+        iban="CH1",
+        bankAccountNumber=BankAccountNumber("ACC1"),
+        bankAccountName=BankAccountName("Test"),
+        totalTaxValue=Decimal("1.2300")
+    )
+    el2 = ba2._build_xml_element(temp_parent)
+    assert el2.get("totalTaxValue") == "1.23"
+    # 2.100 -> 2.1
+    ba3 = BankAccount(
+        iban="CH2",
+        bankAccountNumber=BankAccountNumber("ACC2"),
+        bankAccountName=BankAccountName("Test"),
+        totalTaxValue=Decimal("2.100")
+    )
+    el3 = ba3._build_xml_element(temp_parent)
+    assert el3.get("totalTaxValue") == "2.1"
+    # 5.0 -> 5
+    ba4 = BankAccount(
+        iban="CH3",
+        bankAccountNumber=BankAccountNumber("ACC3"),
+        bankAccountName=BankAccountName("Test"),
+        totalTaxValue=Decimal("5.0")
+    )
+    el4 = ba4._build_xml_element(temp_parent)
+    assert el4.get("totalTaxValue") == "5"
+    # 7.123 -> 7.123 (no change)
+    ba5 = BankAccount(
+        iban="CH4",
+        bankAccountNumber=BankAccountNumber("ACC4"),
+        bankAccountName=BankAccountName("Test"),
+        totalTaxValue=Decimal("7.123")
+    )
+    el5 = ba5._build_xml_element(temp_parent)
+    assert el5.get("totalTaxValue") == "7.123"
 
 def test_optional_boolean_serialization():
     """

--- a/tests/samples/fake_statement.xml
+++ b/tests/samples/fake_statement.xml
@@ -10,10 +10,10 @@
               periodTo="2024-12-31"
               country="CH"
               canton="ZH"
-              totalTaxValue="162750.00"
-              totalGrossRevenueA="4800.00"
+              totalTaxValue="162750"
+              totalGrossRevenueA="4800"
               totalGrossRevenueB="548.29"
-              totalWithHoldingTaxClaim="1680.00"
+              totalWithHoldingTaxClaim="1680"
               minorVersion="22">
   <institution name="Sample Foreign Bank ltd">
     <uid>
@@ -22,28 +22,28 @@
     </uid>
   </institution>
   <client clientNumber="987654" salutation="2" firstName="Max" lastName="Muster"/>
-  <listOfBankAccounts totalTaxValue="35000.00" totalGrossRevenueA="0.00" totalGrossRevenueB="548.29" totalWithHoldingTaxClaim="0.00">
-    <bankAccount bankAccountName="Private Account" bankAccountCountry="LU" bankAccountCurrency="CHF" totalTaxValue="35000.00" totalGrossRevenueA="0.00" totalGrossRevenueB="548.29" totalWithHoldingTaxClaim="0.00" iban="LU9300762011623852957" bankAccountNumber="11623852957">
-      <taxValue referenceDate="2024-12-31" balanceCurrency="CHF" balance="35000.00" value="35000.00"/>
-      <payment paymentDate="2024-06-30" name="Interest Credit" amountCurrency="CHF" amount="500.00" grossRevenueA="0.00" grossRevenueB="500.00" withHoldingTaxClaim="0.00"/>
-      <payment paymentDate="2024-03-15" name="Foreign Interest" amountCurrency="EUR" amount="50.00" exchangeRate="0.9657075" grossRevenueA="0.00" grossRevenueB="48.285375000" withHoldingTaxClaim="0.00"/>
+  <listOfBankAccounts totalTaxValue="35000" totalGrossRevenueA="0" totalGrossRevenueB="548.29" totalWithHoldingTaxClaim="0">
+    <bankAccount bankAccountName="Private Account" bankAccountCountry="LU" bankAccountCurrency="CHF" totalTaxValue="35000" totalGrossRevenueA="0" totalGrossRevenueB="548.29" totalWithHoldingTaxClaim="0" iban="LU9300762011623852957" bankAccountNumber="11623852957">
+      <taxValue referenceDate="2024-12-31" balanceCurrency="CHF" balance="35000" value="35000"/>
+      <payment paymentDate="2024-06-30" name="Interest Credit" amountCurrency="CHF" amount="500" grossRevenueA="0" grossRevenueB="500" withHoldingTaxClaim="0"/>
+      <payment paymentDate="2024-03-15" name="Foreign Interest" amountCurrency="EUR" amount="50" exchangeRate="0.9657075" grossRevenueA="0" grossRevenueB="48.285375" withHoldingTaxClaim="0"/>
     </bankAccount>
   </listOfBankAccounts>
-  <listOfLiabilities totalTaxValue="250000.00" totalGrossRevenueB="3000.00">
-    <liabilityAccount bankAccountName="Mortgage Property ZH" bankAccountCountry="CH" bankAccountCurrency="CHF" totalTaxValue="250000.00" totalGrossRevenueB="3000.00" bankAccountNumber="HYPO-123">
-      <taxValue referenceDate="2024-12-31" balanceCurrency="CHF" balance="250000.00" value="250000.00"/>
-      <payment paymentDate="2024-12-31" name="Mortgage Interest Paid" amountCurrency="CHF" amount="3000.00" grossRevenueB="3000.00"/>
+  <listOfLiabilities totalTaxValue="250000" totalGrossRevenueB="3000">
+    <liabilityAccount bankAccountName="Mortgage Property ZH" bankAccountCountry="CH" bankAccountCurrency="CHF" totalTaxValue="250000" totalGrossRevenueB="3000" bankAccountNumber="HYPO-123">
+      <taxValue referenceDate="2024-12-31" balanceCurrency="CHF" balance="250000" value="250000"/>
+      <payment paymentDate="2024-12-31" name="Mortgage Interest Paid" amountCurrency="CHF" amount="3000" grossRevenueB="3000"/>
     </liabilityAccount>
   </listOfLiabilities>
-  <listOfExpenses totalExpenses="150.00">
-    <expense name="Depot Fees 2024" amountCurrency="CHF" expenses="150.00" expenseType="22" depotNumber="DEPOT-001"/>
+  <listOfExpenses totalExpenses="150">
+    <expense name="Depot Fees 2024" amountCurrency="CHF" expenses="150" expenseType="22" depotNumber="DEPOT-001"/>
   </listOfExpenses>
-  <listOfSecurities totalTaxValue="127750.00" totalGrossRevenueA="4800.00" totalGrossRevenueB="0.00" totalWithHoldingTaxClaim="1680.00" totalLumpSumTaxCredit="0.00" totalNonRecoverableTax="0.00" totalAdditionalWithHoldingTaxUSA="0.00" totalGrossRevenueIUP="0.00" totalGrossRevenueConversion="0.00">
+  <listOfSecurities totalTaxValue="127750" totalGrossRevenueA="4800" totalGrossRevenueB="0" totalWithHoldingTaxClaim="1680" totalLumpSumTaxCredit="0" totalNonRecoverableTax="0" totalAdditionalWithHoldingTaxUSA="0" totalGrossRevenueIUP="0" totalGrossRevenueConversion="0">
     <depot depotNumber="DEPOT-001">
       <security positionId="1" country="CH" currency="CHF" quotationType="PIECE" securityCategory="SHARE" securityName="Roche Holding Genussschein" valorNumber="1203204" isin="CH0012032048">
-        <taxValue referenceDate="2024-12-31" quotationType="PIECE" quantity="500" balanceCurrency="CHF" unitPrice="255.50" value="127750.00" kursliste="1"/>
-        <payment paymentDate="2024-03-18" exDate="2024-03-14" name="Dividend" quotationType="PIECE" quantity="500" amountCurrency="CHF" amountPerUnit="9.6" amount="4800.00" exchangeRate="1" grossRevenueA="4800.00" grossRevenueB="0.00" withHoldingTaxClaim="1680.00" kursliste="1"/>
-        <stock referenceDate="2024-01-01" mutation="0" quotationType="PIECE" quantity="500" balanceCurrency="CHF" value="145000.00"/>
+        <taxValue referenceDate="2024-12-31" quotationType="PIECE" quantity="500" balanceCurrency="CHF" unitPrice="255.5" value="127750" kursliste="1"/>
+        <payment paymentDate="2024-03-18" exDate="2024-03-14" name="Dividend" quotationType="PIECE" quantity="500" amountCurrency="CHF" amountPerUnit="9.6" amount="4800" exchangeRate="1" grossRevenueA="4800" grossRevenueB="0" withHoldingTaxClaim="1680" kursliste="1"/>
+        <stock referenceDate="2024-01-01" mutation="0" quotationType="PIECE" quantity="500" balanceCurrency="CHF" value="145000"/>
       </security>
     </depot>
   </listOfSecurities>

--- a/tests/utils/xml.py
+++ b/tests/utils/xml.py
@@ -1,6 +1,35 @@
 from lxml import etree as ET
 import re
-from typing import Union
+from decimal import Decimal, InvalidOperation
+
+
+_DECIMAL_LEXEME_RE = re.compile(r"^[+-]?(?:\d+\.\d*|\.\d+)$")
+
+
+def _normalize_decimal_lexeme(value: str) -> str:
+    """Normalize decimal lexical forms like 1.20/.30 to 1.2/0.3 for robust XML comparisons."""
+    if not _DECIMAL_LEXEME_RE.match(value):
+        return value
+    try:
+        return format(Decimal(value).normalize(), "f")
+    except (InvalidOperation, ValueError):
+        return value
+
+
+def _normalize_decimal_values(element: ET._Element) -> None:
+    """Normalize decimal-looking attribute and text values recursively."""
+    for attr_name, attr_value in list(element.attrib.items()):
+        element.attrib[attr_name] = _normalize_decimal_lexeme(attr_value)
+
+    if element.text:
+        stripped = element.text.strip()
+        if stripped:
+            normalized = _normalize_decimal_lexeme(stripped)
+            if normalized != stripped:
+                element.text = normalized
+
+    for child in element:
+        _normalize_decimal_values(child)
 
 def normalize_xml(xml_bytes: bytes, remove_xmlns: bool = False) -> str:
     """Normalize XML string by parsing and re-serializing it.
@@ -16,9 +45,9 @@ def normalize_xml(xml_bytes: bytes, remove_xmlns: bool = False) -> str:
     
     # Re-parse and pretty print
     tree = ET.fromstring(normalized_bytes, parser=parser) 
+    _normalize_decimal_values(tree)
     normalized = ET.tostring(tree, pretty_print=True).decode().replace('=".', '="0.')  # type: ignore
     if remove_xmlns:
-        import re
         # Remove xmlns declarations
         normalized = re.sub(r'\s+xmlns(?::[^=]*)?="[^"]*"', '', normalized)
         # Remove schemaLocation and noNamespaceSchemaLocation attributes

